src/dmd/mtype.d

```diff
@@ -5063,6 +5063,10 @@ extern (C++) final class TypeDelegate : TypeNext
 
         version (all)
         {
+            // https://issues.dlang.org/show_bug.cgi?id=20437
+            if (!MODimplicitConv(mod, to.mod))
+                return MATCH.nomatch;
+
```

Comment:
Maybe you should allow head conversion the same way it works for slices. Example:
```d
unittest
{
    shared int[] a;
    f(a);
}

void f(shared(int)[]);
```
Similarly for a delegate it should work like
```d
unittest
{
    shared void delegate() s;
    f(s);
}

void f(void delegate() shared);
```
Since a delegate is also a "fat pointer" just like a slice as per the spec https://dlang.org/spec/abi.html#delegates.

Reply:
If you only check the function types of the delegates then head conversion should work.
```suggestion
            if (!MODimplicitConv(nextOf().mod, to.nextOf().mod))
```

Reply:
This is not obviously a good idea, as `const(void delegate()const)` and `const(void delegate())` must be different types, because all types implicitly convert to const, but a `const(void delegate())`  can't be called. The only reason it would make some sense for `shared` and `immutable` is that there is no way to create a `immutable(void delegate())` or a `shared(void delegate())` that is important to support, but you need to be careful not to break the type system when introducing special cases like this.